### PR TITLE
[BEAM-2290] Fix timestamps for CompressedSource

### PR DESCRIPTION
**Pull request by rfevang**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`.
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by rfevang** [comment]
Hi @dhalperi , can you please take a look?

I'm not sure if the test is the right way to do this, filtering through a DoFn was the only way I saw to get to the timestamp.

**Comment by dhalperi** [comment]
Thanks for catching this!

1. Can you please file a [JIRA issue](https://beam.apache.org/contribute/contribution-guide/#jira-issue-tracker) in the `sdk-java-core` component and update the Pull Request title with the issue number? That way we can track this issue publicly and it will show in the release notes.

2. I will put a few test suggestions, but you did a great start!

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java [hunk 1]
```diff
@@ -632,6 +646,27 @@ protected boolean readNextRecord() throws IOException {
       protected long getCurrentOffset() {
         return offset;
       }
+
+      @Override
+      public Instant getCurrentTimestamp() throws NoSuchElementException {
+        Instant timestamp = ((ByteSource) getCurrentSource()).timestamp;
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: how about we just return `Duration.millis(getCurrentOffset())`? That way elements are timestamped by their offsets fairly directly.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java [hunk 2]
```diff
@@ -632,6 +646,27 @@ protected boolean readNextRecord() throws IOException {
...
+        return timestamp == null ? super.getCurrentTimestamp() : timestamp;
+      }
+    }
+  }
+
+  private static class FilterOnTimestamp extends DoFn<Byte, Byte> {
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: How about `DoFn<Byte, KV<Long, Byte>>` where we output a KV of the element with its timestamp.

Then we can modify the `PAssert` call in the test with to check bytes with indices/timestamps and ensure that every element gets its correct, distinct timestamp.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java (hunk 1 above)
Comment (rfevang) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: Done

File: sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java (hunk 2 above)
Comment (rfevang) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/io/CompressedSourceTest.java]: Done

**Comment by dhalperi** [comment]
Great! Merged, thank you.

**Pull request closed without merging**